### PR TITLE
Reject binary ops where the source is an uninitialized register

### PR DESCRIPTION
**Pull request by Alan-Jowett**
This pull request includes several changes to enhance the handling of uninitialized registers in the assertion extraction logic and adds new test cases to cover these scenarios. The most important changes include updating the `AssertExtractor` class to handle cases where the source register is uninitialized and adding new YAML test cases for various arithmetic and logical operations involving uninitialized registers.

### Enhancements to assertion extraction logic:

* [`src/assertions.cpp`](diffhunk://#diff-6ce2751805a1c8beb6524a37731195a873e763cbeaf8110ba4f60645d65b6facL249-R255): Modified the `AssertExtractor` class to add assertions for source registers when they are uninitialized. (`default` case now checks if the source register is a `Reg` and adds an additional assertion if it is).

### Addition of new test cases:

* [`src/test/test_yaml.cpp`](diffhunk://#diff-d6599eb5fe454ccafc39fb6c17342381a11a931e4cb63345edcdfc7518e87412R42): Added a new YAML test case file `uninit.yaml` to the list of test cases.
* [`test-data/uninit.yaml`](diffhunk://#diff-6a96f9b34eb5f1bb22280c087ac92167475d583c9f48b286f4e3d1539249b631R1-R152): Created a new YAML file with multiple test cases to verify the behavior of arithmetic and logical operations involving uninitialized registers. This includes addition, subtraction, multiplication, division, modulo, AND, OR, and XOR operations.

## Summary by CodeRabbit

- **New Features**
	- Enhanced type checking for binary and atomic operations, ensuring better validation of registers.
	- Introduced a comprehensive YAML test case to validate operations involving uninitialized registers.

- **Bug Fixes**
	- Improved error handling for operations involving uninitialized registers through new test cases.

- **Documentation**
	- Updated comments for clarity in the `AssertExtractor` class.

### CODE REVIEW

File: test-data/uninit.yaml [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+test-case: division of a number and an uninitialized register
+
+pre:
+  - "r0.type=number"
+  - "r0.uvalue=1"
+  - "r0.svalue=1"
+
+code:
+  <start>: |
+    r0 /= r3
```
Comment (dthaler) [inline test-data/uninit.yaml:66]: your yaml test cases are missing some binary operations.
here's the list: ADD, SUB, MUL, DIV, SDIV, OR, AND, LSH, RSH, MOD, SMOD, XOR, ARSH
make sure all of those are tested

File: test-data/uninit.yaml [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+code:
+  <start>: |
+    r0 ^= r3
+
+post:
+  - "r0.type=number"
+
+messages:
+  - "0: Invalid type (r3.type == number)"
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply test-data/uninit.yaml]: 1. Post conditions are still needed. 
2. Error messages match what the verifier produces (can't change them).
3. Error message can't be modified.

File: test-data/uninit.yaml [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+code:
+  <start>: |
+    r0 -= r3
+
+post: []
+
+messages:
+  - "0: Cannot subtract pointers to different regions (r3.type == number or r0.type == r3.type in {ctx, stack, packet})"
+
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply test-data/uninit.yaml:37]: Error messages can't be modified, they are produced by the verifier.

File: test-data/uninit.yaml [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    r0 &= r3
+
+post:
+  - "r0.type=number"
+  - "r0.uvalue=1"
+  - "r0.svalue=1"
+
+messages:
+  - "0: Invalid type (r3.type == number)"
+
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply test-data/uninit.yaml:157]: Post conditions are still needed.

File: test-data/uninit.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply test-data/uninit.yaml:192]: Post conditions are still needed.

File: test-data/uninit.yaml [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+code:
+  <start>: |
+    r0 |= r3
+
+post:
+  - "r0.type=number"
+
+messages:
+  - "0: Invalid type (r3.type == number)"
+
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply test-data/uninit.yaml:175]: Post conditions are still needed.

File: test-data/uninit.yaml [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+post:
+  - "r0.svalue=[0, 1]"
+  - "r0.svalue=r0.uvalue"
+  - "r0.type=number"
+  - "r0.uvalue=[0, 1]"
+
+messages:
+  - "0: Only numbers can be used as divisors (r3 != 0)"
+
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply test-data/uninit.yaml:76]: Post conditions are still needed.

File: test-data/uninit.yaml (hunk 1 above)
Comment (Alan-Jowett) [reply test-data/uninit.yaml:66]: I think I added all of them, please let me know if I missed anything.

File: test-data/uninit.yaml [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+  <start>: |
+    r0 s/= r3
+
+post:
+  - "r0.svalue=[-1, 1]"
+  - "r0.type=number"
+
+messages:
+  - "0: Only numbers can be used as divisors (r3 != 0)"
+
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply test-data/uninit.yaml:95]: Error message come from the verifier and can't be modified.

File: test-data/uninit.yaml [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+code:
+  <start>: |
+    r0 <<= r3
+
+post:
+  - "r0.type=number"
+
+messages:
+  - "0: Invalid type (r3.type == number)"
+
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply test-data/uninit.yaml:211]: Post conditions are still needed.

File: test-data/uninit.yaml [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+code:
+  <start>: |
+    r0 s>>= r3
+
+post:
+  - "r0.type=number"
+
+messages:
+  - "0: Invalid type (r3.type == number)"
+
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply test-data/uninit.yaml:247]: Post conditions are still needed.

File: src/test/test_yaml.cpp [hunk 10]
```diff
@@ -39,5 +39,6 @@ YAML_CASE("test-data/sext.yaml")
 YAML_CASE("test-data/shift.yaml")
 YAML_CASE("test-data/stack.yaml")
 YAML_CASE("test-data/subtract.yaml")
+YAML_CASE("test-data/uninit.yaml")
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply src/test/test_yaml.cpp:42]: That doesn't make sense. Alphabetical order would put i before o.

**Review by dthaler — APPROVED** [review]

File: src/assertions.cpp [hunk 11]
```diff
@@ -196,6 +196,7 @@ class AssertExtractor {
         res.emplace_back(TypeConstraint{ins.access.basereg, TypeGroup::pointer});
         res.emplace_back(
             ValidAccess{ins.access.basereg, ins.access.offset, Imm{static_cast<uint32_t>(ins.access.width)}, false});
+        res.emplace_back(TypeConstraint{ins.valreg, TypeGroup::number});
```
[the remark this answers is not in the document]
Comment (elazarg) [reply src/assertions.cpp:199]: @Alan-Jowett I may be confused, but isn't coderabbit correct here?

[the remark this answers is not in the document]
Comment (elazarg) [reply src/assertions.cpp:199]: The implementation should be:
```cpp
    vector<Assert> operator()(const Atomic& ins) const {
        return {
            Assert{TypeConstraint{ins.valreg, TypeGroup::number}},
            Assert{TypeConstraint{ins.access.basereg, TypeGroup::pointer}},
            Assert{ValidAccess{ins.access.basereg, ins.access.offset, Imm{static_cast<uint32_t>(ins.access.width)}, false}},
        }
    }

[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply src/assertions.cpp:199]: Agreed. Will make that change and post an update shortly. Was too quick to dismiss coderabbit.

**Comment by Alan-Jowett** [comment]
Applied feedback and rebased to minimize commits.

File: src/assertions.cpp [hunk 12]
```diff
@@ -193,15 +193,10 @@ class AssertExtractor {
 
     vector<Assert> operator()(const Atomic& ins) const {
         vector<Assert> res;
+        res.emplace_back(TypeConstraint{ins.valreg, TypeGroup::number});
```
[the remark this answers is not in the document]
Comment (elazarg) [reply src/assertions.cpp:196]: I agree

**Pull request merged**